### PR TITLE
[SPARK-11520][ML] RegressionMetrics should support instance weights

### PR DESCRIPTION
**Pull request by Lewuathe**
This will be important to improve LinearRegressionSummary, which currently has a mix of weighted and unweighted metrics.

### CODE REVIEW

**Comment by Lewuathe** [comment]
@jkbradley Could you review this? Thanks.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala [hunk 1]
```diff
@@ -39,16 +51,17 @@ class RegressionMetrics @Since("1.2.0") (
...
    */
   private lazy val summary: MultivariateStatisticalSummary = {
-    val summary: MultivariateStatisticalSummary = predictionAndObservations.map {
-      case (prediction, observation) => Vectors.dense(observation, observation - prediction)
+    val summary: MultivariateStatisticalSummary = predictionAndObservationsWithWeight.map {
+      case (prediction, observation, weight) =>
+        (Vectors.dense(observation, observation - prediction), weight)
     }.aggregate(new MultivariateOnlineSummarizer())(
-        (summary, v) => summary.add(v),
+        (summary, sample) => summary.add(sample._1, sample._2),
```
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: We should rename this `summary` variable, it is used 3 times for different objects.

Also, `sample` is a bit too generic here, and using the `_xx` methods are not very intuitive. I suggest you unpack the tuple fully: `{ case (currentSummary, (vec, weight)) => currentSummary.add(vec, weight) }`

File: mllib/src/test/scala/org/apache/spark/mllib/evaluation/RegressionMetricsSuite.scala [hunk 2]
```diff
@@ -109,4 +109,55 @@ class RegressionMetricsSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+    *             (observations, observations - predictions)
+    * mean        (1.7, 0.05)
+    * variance    (7.3, 0.3)
+    * numNonZeros (0.5, 0.5)
+    * max         (7.0, 0.75)
+    * min         (-0.5, -0.75)
+    * normL2      (2.0, 0.32596)
+    * normL1      (1.05, 0.2)
+    *
+    * explainedVariance: sum((preds - 1.7)^2) / count = 10.1775
```
Comment (thunterdb) [inline mllib/src/test/scala/org/apache/spark/mllib/evaluation/RegressionMetricsSuite.scala]: I am having an issue with these formulas, based on the very wikipedia page you are referencing. It should not be normalized by the count but by the weight sum, right?

File: mllib/src/test/scala/org/apache/spark/mllib/evaluation/RegressionMetricsSuite.scala [hunk 3]
```diff
@@ -109,4 +109,55 @@ class RegressionMetricsSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+    * meanAbsoluteError: normL1(1) / count = 0.05
+    * meanSquaredError: normL2(1)^2 / count = 0.02656
+    * rootMeanSquaredError: sqrt(meanSquaredError) = 0.16298
+    * r2: 1 - normL2(1)^2 / (variance(0) * (count - 1)) = 0.9951484
```
Comment (thunterdb) [inline mllib/src/test/scala/org/apache/spark/mllib/evaluation/RegressionMetricsSuite.scala]: Same thing for this formula.

**Comment by thunterdb** [comment]
@Lewuathe thanks for your patch. I think it will require more work in `RegressionMetrics` to fully implement weighted metrics. We need to do the following changes:
- expose `weightSum` in `MultivariateStatisticalSummary` (as a developer API)
- the computations of `SSreg` and `SStot` should take the weights into account
- in `RegressionMetrics`, all references to `summary.count` should be replaced by `summary.weightSum` 
Given that the default weights are 0, it should give the same result.

**Pull request closed without merging**